### PR TITLE
v0 Datafusion with late materialization

### PR DESCRIPTION
**Pull request by a10y**
Unfinished, just opening this as I continue to get things working.

This PR augments the original Vortex connection for Datafusion, with an implementation of filter pushdown that allows us to perform late materialization on as many columns as possible.

Pushdown support will be able to get flagged on/off so we can run benchmarks testing different strategies.

I'm hoping to have an initial version of this with a benchmark harness tonight.

### CODE REVIEW

File: bench-vortex/benches/datafusion_benchmark.rs [hunk 1]
```diff
@@ -0,0 +1,160 @@
...
+    let mut group = c.benchmark_group("datafusion");
+
+    let session = SessionContext::new();
+
+    let arrow_dataset = toy_dataset_arrow();
+    let arrow_table =
+        Arc::new(MemTable::try_new(arrow_dataset.schema(), vec![vec![arrow_dataset]]).unwrap());
+
+    group.bench_function("arrow", |b| {
+        let arrow_table = arrow_table.clone();
```
Comment (robert3005) [inline bench-vortex/benches/datafusion_benchmark.rs]: You don't need these clones

File: vortex-array/src/array/bool/compute/compare.rs [hunk 2]
```diff
@@ -8,6 +8,7 @@ use crate::compute::compare::CompareFn;
 use crate::{Array, ArrayTrait, IntoArray, IntoArrayVariant};
 
 impl CompareFn for BoolArray {
+    // TODO(aduffy): replace these with Arrow compute kernels.
```
Comment (robert3005) [inline vortex-array/src/array/bool/compute/compare.rs:11]: makes sense for primitives... and we don't have any other impls yet

File: vortex-datafusion/src/lib.rs [hunk 3]
```diff
@@ -99,10 +132,29 @@ impl TableProvider for VortexInMemoryTableProvider {
...
+        let filter_exprs: Option<Vec<Expr>> = if filters.is_empty() {
+            None
+        } else {
+            Some(filters.to_vec())
+        };
+
+        let filter_projection = filter_exprs
+            .clone()
+            .map(|exprs| get_filter_projection(exprs.as_slice(), self.schema_ref.clone()));
+
```
Comment (robert3005) [inline vortex-datafusion/src/lib.rs]: this is a bit redundant? `&[Expr] -> Vec<Expr> -> Option<Vec<Expr>> -> Clone -> &[Expr]`. It works just fine if you keep everything as `&[Expr]`

File: vortex-datafusion/src/plans.rs [hunk 4]
```diff
@@ -0,0 +1,477 @@
...
+    ) -> DFResult<SendableRecordBatchStream> {
+        assert_eq!(
+            partition, 0,
+            "single partitioning only supported by TakeOperator"
+        );
+
+        let stream_schema = Arc::new(infer_schema(self.filter_struct.dtype()));
+
+        let filter_struct = self.filter_struct.clone();
+        let inner = Box::pin(async move {
```
Comment (robert3005) [inline vortex-datafusion/src/plans.rs]: we likely would like to avoid converting to arrow here but that can come later

File: vortex-datafusion/src/plans.rs [hunk 5]
```diff
@@ -0,0 +1,477 @@
...
+        // We want to avoid ever calling it again.
+        let record_batch = ready!(this.inner.poll(cx))?;
+        *this.polled_inner = true;
+
+        // Using a local SessionContext, generate a physical plan to execute the conjunction query
+        // against the filter columns.
+        //
+        // The result of a conjunction expression is a BooleanArray containing `true` for rows
+        // where the conjunction was satisfied, and `false` otherwise.
+        let session = SessionContext::new();
```
Comment (robert3005) [inline vortex-datafusion/src/plans.rs]: Is there a way to pass context into this? I recall this function being slow

File: vortex-datafusion/src/plans.rs [hunk 6]
```diff
@@ -0,0 +1,477 @@
...
+        let df_schema = DFSchema::try_from(this.schema_ref.clone())?;
+        let physical_expr =
+            session.create_physical_expr(this.conjunction_expr.clone(), &df_schema)?;
+        let selection = physical_expr
+            .evaluate(&record_batch)?
+            .into_array(record_batch.num_rows())?;
+
+        // Convert the `selection` BooleanArray into a UInt64Array of indices.
+        let selection_indices: Vec<u64> = selection
```
Comment (robert3005) [inline vortex-datafusion/src/plans.rs:197]: there should be a compute function that takes with a bitmask... `mask`?

File: vortex-datafusion/src/plans.rs [hunk 7]
```diff
@@ -0,0 +1,477 @@
...
+                vec![],
+                &opts,
+            )
+            .unwrap())));
+        }
+
+        // Assemble the output columns using the row indices.
+        // NOTE(aduffy): this re-decodes the fields from the filter schema, which is unnecessary.
+        let mut columns = Vec::new();
+        for (output_idx, src_idx) in this.output_projection.iter().enumerate() {
```
Comment (robert3005) [inline vortex-datafusion/src/plans.rs]: we should project the array before and then just do a take here

File: vortex-datafusion/src/lib.rs [hunk 8]
```diff
@@ -115,87 +167,250 @@ impl TableProvider for VortexInMemoryTableProvider {
...
+        &struct_array,
+    ))
+}
+
+/// Check if the given expression tree can be pushed down into the scan.
+fn can_be_pushed_down(expr: &Expr, schema_columns: &HashSet<String>) -> DFResult<bool> {
+    // If the filter references a column not known to our schema, we reject the filter for pushdown.
+    // TODO(aduffy): is this necessary? Under what conditions would this happen?
+    let column_refs = get_column_references(expr);
+    if !column_refs.is_subset(schema_columns) {
```
Comment (robert3005) [inline vortex-datafusion/src/lib.rs]: not sure how datafusion produces the output schema but if you have `SELECT A WHERE A > B` depending on the internal implementation it might not have `B` in the output schema

File: vortex-datafusion/src/lib.rs [hunk 9]
```diff
@@ -115,87 +167,250 @@ impl TableProvider for VortexInMemoryTableProvider {
...
-            array: self.array.clone(),
-            projection: projection.cloned(),
-            plan_properties,
-        }))
+        let output_projection: Vec<usize> = match projection {
+            None => (0..self.schema_ref.fields().len()).collect(),
+            Some(proj) => proj.clone(),
+        };
+
+        match (filter_exprs, filter_projection) {
```
Comment (robert3005) [inline vortex-datafusion/src/lib.rs]: there should be a better way of representing this. You could push filter projection construction into the match

**Review by robert3005 — COMMENTED** [review]
I know it's unfinished so I tried to leave high level comments only

File: vortex-datafusion/src/plans.rs (hunk 5 above)
Comment (a10y) [reply vortex-datafusion/src/plans.rs]: yea i don't like this either, gonna find a better way to do this before we merge

File: vortex-datafusion/src/plans.rs (hunk 6 above)
Comment (a10y) [reply vortex-datafusion/src/plans.rs:197]: i didn't see one for Vortex

File: vortex-datafusion/src/lib.rs (hunk 8 above)
Comment (a10y) [reply vortex-datafusion/src/lib.rs]: DF differentiates between the table schema and the output projection, and filters can reference things that are not in the output projection but are in the table schema. They explain this behavior in https://docs.rs/datafusion/latest/datafusion/datasource/provider/trait.TableProvider.html#tymethod.scan under the `Filters` section.

The thing I was more concerned about is I don't know if it can pushdown filters that reference columns in other tables, like broadcast joins. Maybe this is too defensive

File: vortex-datafusion/src/plans.rs (hunk 6 above)
Comment (robert3005) [reply vortex-datafusion/src/plans.rs:197]: Indeed, it was more of a statement that one like that should exist

File: bench-vortex/benches/datafusion_benchmark.rs (hunk 1 above)
Comment (a10y) [reply bench-vortex/benches/datafusion_benchmark.rs]: removed

File: vortex-datafusion/src/plans.rs (hunk 5 above)
Comment (a10y) [reply vortex-datafusion/src/plans.rs]: it turns out they have a naked function you can just call directly

File: vortex-datafusion/src/lib.rs (hunk 9 above)
Comment (a10y) [reply vortex-datafusion/src/lib.rs]: pushed the projection into the match arm

File: vortex-datafusion/src/lib.rs (hunk 3 above)
Comment (a10y) [reply vortex-datafusion/src/lib.rs]: fair, fixed

**Comment by a10y** [comment]
Output of the `datafusion_benchmark` on my MBP.

Note that `vortex-nopushdown-uncompressed` should actually be `vortex-nopushdown-compressed`, and `vortex-nopushdown-uncompressed #2` is the actual `vortex-nopushdown-uncompressed`.

image

Even though this is synthetic data, it still illustrates that decoding overhead is the driving factor in execution time.

There's also some latency between uncompressed Vortex with no pushdown and Arrow with no pushdown, but that time is roughly the ~130µs it takes to do the Vortex -> Arrow conversion (benchmarked that separately, not in the repo).

**Comment by robert3005** [comment]
Right now we don't run the filters on compressed data which would probably be the thing to fix. Anyway, this seems fixable

**Comment by a10y** [comment]
I agree. I'm going to address the last few comments of your original review and then convert this to "Ready for review"

File: vortex-datafusion/src/plans.rs (hunk 7 above)
Comment (a10y) [reply vortex-datafusion/src/plans.rs]: done

File: vortex-datafusion/src/plans.rs (hunk 4 above)
Comment (a10y) [reply vortex-datafusion/src/plans.rs]: I changed this to be a future that returns vortex array, so we can apply some vortex-level filtering in the operator

File: vortex-array/src/stats/statsset.rs [hunk 10]
```diff
@@ -27,6 +27,36 @@ impl StatsSet {
         }
     }
 
+    // pub fn constant(len: usize, scalar: &Scalar) -> Self {
```
Comment (robert3005) [inline vortex-array/src/stats/statsset.rs]: I think you want to remove this? Or is this the thing you want?

Comment (a10y) [reply vortex-array/src/stats/statsset.rs]: i'll remove it from this PR and add it back later, it will be more useful when we want to do stats-based pushdown to have this

File: vortex-datafusion/src/lib.rs [hunk 11]
```diff
@@ -94,108 +127,297 @@ impl TableProvider for VortexInMemoryTableProvider {
...
+                .sorted()
+                .collect();
+
+            projection
         }
 
+        let filter_exprs: Option<Vec<Expr>> = if filters.is_empty() {
+            None
+        } else {
+            Some(filters.to_vec())
```
Comment (robert3005) [inline vortex-datafusion/src/lib.rs]: my earlier comment meant that you can get rid of this `to_vec`

File: pyvortex/test/test_array.py [hunk 12]
```diff
@@ -18,10 +18,9 @@ def test_varbin_array_round_trip():
 
 def test_varbin_array_take():
     a = vortex.encode(pa.array(["a", "b", "c", "d"]))
-    # TODO(ngates): ensure we correctly round-trip to a string and not large_string
```
Comment (a10y) [inline pyvortex/test/test_array.py:21]: todo -> todone

File: vortex-array/src/array/varbin/compute/take.rs [hunk 13]
```diff
@@ -49,7 +49,7 @@ fn take<I: NativePType, O: NativePType>(
         return Ok(take_nullable(dtype, offsets, data, indices, v));
     }
 
-    let mut builder = VarBinBuilder::<I>::with_capacity(indices.len());
+    let mut builder = VarBinBuilder::<O>::with_capacity(indices.len());
```
Comment (a10y) [inline vortex-array/src/array/varbin/compute/take.rs:52]: previously, take(VarBinArray) was yielding a new VarBinArray with offset that had ptype `indices.ptype`, we actually wanted them to maintain their original ptype.

This is also why the arrow representation of string arrays often switched from Utf8 -> LargeUtf8 after doing a take, as in the pytest above

File: vortex-datafusion/src/lib.rs [hunk 14]
```diff
@@ -261,7 +483,7 @@ impl ExecutionPlan for VortexMemoryExec {
             self.array.clone()
         };
 
-        Self::execute_single_chunk(chunk, &self.projection, context)
+        execute_unfiltered(&chunk, &self.scan_projection)
```
Comment (robert3005) [inline vortex-datafusion/src/lib.rs]: not sure why this change, you immediately clone the chunk in the function. How about we also change the scan_projection to `&[usize]`?

**Review by robert3005 — APPROVED** [review]
There's some style nits to avoid clones. Should figure out if we really need to validate filter column references

File: vortex-datafusion/src/lib.rs (hunk 14 above)
Comment (a10y) [reply vortex-datafusion/src/lib.rs]: good point

File: vortex-datafusion/src/lib.rs [hunk 15]
```diff
@@ -94,108 +127,297 @@ impl TableProvider for VortexInMemoryTableProvider {
...
+        &output_projection,
+        row_selector_op.clone(),
+        &struct_array,
+    ))
+}
+
+/// Check if the given expression tree can be pushed down into the scan.
+fn can_be_pushed_down(expr: &Expr, schema_columns: &HashSet<String>) -> DFResult<bool> {
+    // If the filter references a column not known to our schema, we reject the filter for pushdown.
+    // TODO(aduffy): is this necessary? Under what conditions would this happen?
```
Comment (robert3005) [inline vortex-datafusion/src/lib.rs]: Reading the docs I don't think this check is necessary. TableProvider returns the schema so datafusion will know if the query is valid or not

Comment (a10y) [reply vortex-datafusion/src/lib.rs]: i just removed it, if we end up needing to add it back later for some reason we can reference this PR

File: vortex-datafusion/src/lib.rs (hunk 11 above)
Comment (a10y) [reply vortex-datafusion/src/lib.rs]: done
